frame/system/module.go

```diff
@@ -867,3 +1022,127 @@ func mutateAccount(account *primitives.AccountInfo, data *primitives.AccountData
+// `OnSetCode` configuration of the runtime.
+func (m module) DoApplyAuthorizeUpgrade(codeBlob sc.Sequence[sc.U8]) (primitives.PostDispatchInfo, error) {
+	_, err := m.validateAuthorizedUpgrade(codeBlob)
+	if err != nil {
+		return primitives.PostDispatchInfo{}, err
+	}
+
+	err = m.OnSetCode.SetCode(codeBlob)
+	if err != nil {
+		return primitives.PostDispatchInfo{}, err
+	}
+
+	m.storage.AuthorizedUpgrade.Clear()
+
+	post := primitives.PostDispatchInfo{
+		// consume the rest of the block to prevent further transactions
+		ActualWeight: sc.NewOption[primitives.Weight](m.constants.BlockWeights.MaxBlock),
+		// no fee for valid upgrade
+		PaysFee: primitives.PaysNo,
+	}
+
+	return post, nil
+}
+
+// Check that provided `code` can be upgraded to. Namely, check that its hash matches an
+// existing authorization and that it meets the specification requirements of `can_set_code`.
+func (m module) validateAuthorizedUpgrade(codeBlob sc.Sequence[sc.U8]) (primitives.H256, error) {
+	authorization, err := m.storage.AuthorizedUpgrade.Get()
+	if err != nil {
+		return primitives.H256{}, NewDispatchErrorNothingAuthorized(m.Index)
+	}
+
+	hash := m.ioHashing.Blake256(sc.SequenceU8ToBytes(codeBlob))
+	actualHash, err := primitives.NewH256(sc.BytesToFixedSequenceU8(hash)...)
+	if err != nil {
+		return primitives.H256{}, err
+	}
+
+	if !reflect.DeepEqual(actualHash, authorization.CodeHash) {
+		return primitives.H256{}, NewDispatchErrorUnauthorized(m.Index)
+	}
+
+	if authorization.CheckVersion {
+		err := m.CanSetCode(codeBlob)
+		if err != nil {
+			return primitives.H256{}, err
+		}
+	}
+
+	return actualHash, nil
+}
+
+// EnsureRoot ensures that the origin represents the root.
+func EnsureRoot(origin primitives.RuntimeOrigin) error {
+	if origin.IsRootOrigin() {
+		return nil
+	} else {
+		return primitives.NewDispatchErrorBadOrigin()
+	}
+}
+
+// EnsureSigned ensures that the origin represents a signed extrinsic (i.e. transaction).
+// Returns `Ok` with the account that signed the extrinsic or an `Err` otherwise.
+func EnsureSigned(origin primitives.RawOrigin) (sc.Option[primitives.AccountId], error) {
+	if origin.IsSignedOrigin() {
+		return sc.NewOption[primitives.AccountId](origin.VaryingData[1]), nil
+	}
+
+	return sc.Option[primitives.AccountId]{}, primitives.NewDispatchErrorBadOrigin()
+}
+
+// EnsureSignedOrRoot ensures the origin represents either a signed extrinsic or the root.
+// Returns an empty Option if the origin is `Root`.
+// Returns an Option with the signer if the origin is signed.
+// Returns a `BadOrigin` error if neither of the above.
+func EnsureSignedOrRoot(origin primitives.RawOrigin) (sc.Option[primitives.AccountId], error) {
+	if origin.IsRootOrigin() {
+		return sc.NewOption[primitives.AccountId](nil), nil
+	} else if origin.IsSignedOrigin() {
+		return sc.NewOption[primitives.AccountId](origin.VaryingData[1]), nil
+	}
+
+	return sc.Option[primitives.AccountId]{}, primitives.NewDispatchErrorBadOrigin()
+}
```

Comment:
1000+ lines, perhaps we could split into multiple files, for starters the metadata?